### DISCUSSION TITLE
How would I flow a parent Trace through a queued message using RabbitMQ?

### DISCUSSION THREAD

**Discussion by ghost** [datalust/seq-tickets#2096] [Help]
I have two applications (producer and consumer) and I would like to be able to visualize a trace that includes spans that originate from the producer app and flow through to completion in the consumer application. 

I'm assuming I need to add the trace information to the message properties when it's queued and then extract/parse them and "reattach" them to the activity in the consumer but I'm not too sure. 

I am using SerilogTracing and not the OpenTelemetry .NET SDK

Any help/reference/advice is most appreciated.

**Comment by nblumhardt** [comment]
Hi @wes-mytrailmate, thanks for your message.

The trick is to grab `Activity.Current.Id` at the point your producer application creates a message, and include this in some metadata on that message to use as the parent of a new child activity at the consumer. The activity id, as long as you have `IdFormat` set to `W3C`, will be essentially the format of the HTTP `traceparent` header used in client/server trace propagation.

There isn't a standardized way to do this with RabbitMQ just yet, ([something seems to be in the works at the AMQP level](https://w3c.github.io/trace-context-amqp/#traceparent-amqp-format]), but attaching the id as a message header, and reading it again in some kind of hook on the consumer side before handling the message, should be enough.

There's some plumbing involved - if you dig in and hit any obstacles, let me know, happy to take a look and help where I can.

**Comment by wazzamatazz** [comment]
FYI it looks like version 7 of the .NET RabbitMQ client will have OpenTelemetry trace support: https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528
